File: Code/Supply-Chain-Management-main dekhte hobe/supply-chain-management - angular/src/app/inventory/raw-material-category/raw-material-category.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ApiResponse } from '../../util/api.response';
import { RawMaterialCategory } from './model/raw-material-category.model';

@Injectable({
  providedIn: 'root'
})
export class RawMaterialCategoryService {

  private apiUrl = 'http://localhost:8080/api/rawmaterialcategory';

  constructor(private http: HttpClient) { }

  getAllRawMaterialCategories(): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/list`);
  }

  saveRawMaterialCategory(rawMaterialCategory: RawMaterialCategory): Observable<ApiResponse> {
    return this.http.post<ApiResponse>(`${this.apiUrl}/save`, rawMaterialCategory);
  }

  updateRawMaterialCategory(rawMaterialCategory: RawMaterialCategory): Observable<ApiResponse> {
    return this.http.put<ApiResponse>(`${this.apiUrl}/update`, rawMaterialCategory);
  }

  deleteRawMaterialCategoryById(id: number): Observable<ApiResponse> {
    return this.http.delete<ApiResponse>(`${this.apiUrl}/delete/${id}`);
  }

  findRawMaterialCategoryById(id: number): Observable<ApiResponse> {
    return this.http.get<ApiResponse>(`${this.apiUrl}/${id}`);
  }

}

File: Code/Supply-Chain-Management-main dekhte hobe/SCMSystem_SpringBoot/src/main/java/com/shamima/SCMSystem/goods/service/RawMaterialCategoryService.java
package com.shamima.SCMSystem.goods.service;

import com.shamima.SCMSystem.goods.entity.RawMaterial;
import com.shamima.SCMSystem.goods.entity.RawMaterialCategory;
import com.shamima.SCMSystem.goods.repository.RawMaterialCategoryRepository;
import com.shamima.SCMSystem.goods.repository.RawMaterialRepository;
import com.shamima.SCMSystem.util.ApiResponse;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class RawMaterialCategoryService {

    @Autowired
    private RawMaterialCategoryRepository rawMaterialCategoryRepository;

    @Autowired
    private RawMaterialRepository rawMaterialRepository;

    public ApiResponse getRawMaterialCategories() {
        ApiResponse apiResponse = new ApiResponse(false);
        try {
            List<RawMaterialCategory> categories = rawMaterialCategoryRepository.findAll();
            apiResponse.setSuccess(true);
            apiResponse.setData("categories", categories);
        } catch (Exception e) {
            apiResponse.setMessage(e.getMessage());
        }
        return apiResponse;
    }

    @Transactional
    public ApiResponse saveRawMaterialCategory(RawMaterialCategory rmc) {
        ApiResponse apiResponse = new ApiResponse(false);
        try {
            rawMaterialCategoryRepository.save(rmc);
            apiResponse.setSuccess(true);
            apiResponse.setMessage("Raw Material Category saved successfully");
            apiResponse.setData("category", rmc);
        } catch (Exception e) {
            apiResponse.setMessage(e.getMessage());
        }
        return apiResponse;
    }

    @Transactional
    public ApiResponse updateRawMaterialCategory(RawMaterialCategory rmc) {
        ApiResponse apiResponse = new ApiResponse(false);
        try {
            RawMaterialCategory existingRMC = rawMaterialCategoryRepository.findById(rmc.getId()).orElse(null);
            if (existingRMC == null) {
                apiResponse.setMessage("Category with ID " + rmc.getId() + " not found");
                return apiResponse;
            }
            rawMaterialCategoryRepository.save(rmc);
            apiResponse.setSuccess(true);
            apiResponse.setMessage("Raw Material Category updated successfully");
            apiResponse.setData("category", rmc);
        } catch (Exception e) {
            apiResponse.setMessage(e.getMessage());
        }
        return apiResponse;
    }

    public ApiResponse findRawMaterialCategoryById(long id) {
        ApiResponse apiResponse = new ApiResponse(false);
        try {
            RawMaterialCategory category = rawMaterialCategoryRepository.findById(id).orElse(null);
            if (category == null) {
                apiResponse.setMessage("Category with ID " + id + " not found");
                return apiResponse;
            }
            apiResponse.setSuccess(true);
            apiResponse.setData("category", category);
        } catch (Exception e) {
            apiResponse.setMessage(e.getMessage());
        }
        return apiResponse;
    }

    public ApiResponse deleteRawMaterialCategoryById(long id) {
        ApiResponse apiResponse = new ApiResponse(false);
        try {
            List<RawMaterial> existingRawMaterials = rawMaterialRepository.findAllByCategoryId(id);
            for (RawMaterial rawMaterial : existingRawMaterials) {
                rawMaterialRepository.delete(rawMaterial);
            }
            rawMaterialCategoryRepository.deleteById(id);
            apiResponse.setSuccess(true);
            apiResponse.setMessage("Raw Material Category and related materials deleted successfully");
        } catch (Exception e) {
            apiResponse.setMessage(e.getMessage());
        }
        return apiResponse;
    }
}

